desc/objectives/_free_boundary.py

```diff
@@ -754,6 +755,338 @@ def _print(fmt, fmax, fmin, fmean, norm, units):
+                Bplasma=Bplasma,
+            )
+
+        timer.stop("Precomputing transforms")
+        if verbose > 1:
+            timer.disp("Precomputing transforms")
+
+        self._dim_f = eval_grid.num_nodes
+
+        if self._normalize:
+            scales = compute_scaling_factors(eq)
+            self._normalization = scales["B"]
+
+        super().build(use_jit=use_jit, verbose=verbose)
+
+    def compute(self, params_1=None, params_2=None, constants=None):
+        """Compute boundary force error.
+
+        Parameters
+        ----------
+        params_1, params_2 : dict
+            Dictionary of the external field's or equilibrium's degrees of freedom.
+            If both the equilibrium and field are not fixed then `params_1` is the
+            field params and `params_2` is the eq params.
+        constants : dict
+            Dictionary of constant data, eg transforms, profiles etc. Defaults to
+            self.constants
+
+        Returns
+        -------
+        f : ndarray
+            Bnorm from Bext and Bplasma
+
+        """
+        if constants is None:
+            constants = self.constants
+
+        # get params based on the order of what is in `things`
+        if self._eq_fixed:
+            field_params = params_1
+        elif self._field_fixed:
+            eq_params = params_1
+            field_params = self._ext_field.params_dict
+        else:
+            field_params = params_1
+            eq_params = params_2
+
+        # Now, calculate Bplasma and Bext
+        if self._eq_fixed or self._vacuum:
+            eval_data = constants["eval_data"]
+            Bplasma = constants["Bplasma"]
+        else:
+            eval_data = compute_fun(
+                "desc.equilibrium.equilibrium.Equilibrium",
+                self._data_keys,
+                params=eq_params,
+                transforms=constants["eval_transforms"],
+                profiles=constants["eval_profiles"],
+            )
+
+            src_data = compute_fun(
+                "desc.equilibrium.equilibrium.Equilibrium",
+                self._data_keys,
+                params=eq_params,
+                transforms=constants["src_transforms"],
+                profiles=constants["src_profiles"],
+            )
+
+            # don't need extra B/2 since we only care about normal component
+            Bplasma = -singular_integral(
+                eval_data,
+                src_data,
+                "biot_savart",
+                constants["interpolator"],
+            )
+
+        x = jnp.array(
+            [
+                eval_data["R"],
+                eval_data["zeta"],
+                eval_data["Z"],
+            ]
+        ).T
+
+        # can't pre-compute Bext because it is dependent on eval_grid
+        Bext = self._ext_field.compute_magnetic_field(
+            x, grid=self._field_grid, basis="rpz", params=field_params
+        )
+
+        f = jnp.sum((Bext + Bplasma) * eval_data["n_rho"], axis=-1)
```

Comment:
do we want to weight this by the area element?

Reply:
I'm not sure what this means

Reply:
basically right now this is an approximation to in the integral 

$$\int B\cdot n d\theta d\zeta$$

which has units of Tesla.

Another option is to have it approximate the integral

$$\int B\cdot n dA = \int B\cdot n |e_\theta \times e_\zeta| d\theta d\zeta$$

with units of T*m^2, where $|e_\theta \times e_\zeta|$ is the area element that accounts for the fact that different chunks of $\theta, \zeta$ cover different areas in real space.

I'm not sure which is more conventional/correct, @dpanici might know more.

Reply:
I think it should include the area element, yes. I just checked and simsopt defines it that way too.

Reply:
Okay thank you. So would the code look something like `f = jnp.sum((Bext + Bplasma) * eval_data["n_rho"] * eval_data["S(r)"], axis=-1)`?

Reply:
also do I need to do anything else when comparing with the computed Bnorm? i.e. `field.compute_Bnormal`

Reply:
it should be the local area element `"|e_theta x e_zeta|"`:

```python
f = jnp.sum((Bext + Bplasma) * eval_data["n_rho"], axis=-1) * eval_data["|e_theta x e_zeta|"]